server/src/main/java/io/whitefox/core/EvalHelper.java

```diff
@@ -0,0 +1,103 @@
+package io.whitefox.core;
+
+import java.sql.Timestamp;
+import java.util.List;
+import java.util.Objects;
+
+public class EvalHelper {
+
+        // Implements "equal" between two leaf operations.
+        static Boolean equal(List<LeafOp> children, EvalContext ctx) {
+          var leftChild = children.get(0);
+          var leftType = leftChild.evalExpectValueAndType(ctx).getRight();
+          var leftVal = leftChild.evalExpectValueAndType(ctx).getLeft();
+
+          var rightChild = children.get(1);
+          var rightType = rightChild.evalExpectValueAndType(ctx).getRight();
+          var rightVal = rightChild.evalExpectValueAndType(ctx).getLeft();
+          // If the types don't match, it implies a malformed predicate tree.
+          // We simply throw an exception, which will cause filtering to be skipped.
+          if (!Objects.equals(leftType, rightType)) {
+              throw new IllegalArgumentException(
+              "Type mismatch: " + leftType + " vs " + rightType + " for " +
+                  leftChild + " and " + rightChild
+              );
+          }
+
+          // We throw an exception for nulls, which will skip filtering.
+          if (leftVal == null || rightVal == null) {
+              throw new IllegalArgumentException(
+              "Comparison with null is not supported: " + leftChild + " and " + rightChild
+              );
+          }
+
+          switch (leftType) {
+                  case OpDataTypes.BoolType: return Boolean.valueOf(leftVal) == Boolean.valueOf(rightVal);
+                  case OpDataTypes.IntType: return Integer.parseInt(leftVal) == Integer.parseInt(rightVal);
+                  case OpDataTypes.LongType: return Long.parseLong(leftVal) == Long.parseLong(rightVal);
+                  case OpDataTypes.StringType: return leftVal.equals(rightVal);
+                  case OpDataTypes.DateType:
+                  return java.sql.Date.valueOf(leftVal).equals(java.sql.Date.valueOf(rightVal));
```

Comment:
does this comply with the way delta writes out dates?

Reply:
yep the whole implementation is more or less based on the delta-sharing one, adapted to Java

Reply:
to make this more clear anyway i would factor out this to a dedicated utility class